### PR TITLE
#30983 Ensure capacity returned of HashMap is max(capacity, length).

### PR DESCRIPTION
**Pull request by rthomas**
Using the test @bluss suggested in #30983

### CODE REVIEW

**Comment by bluss** [comment]
@rthomas Welcome to Rust! It's great that you want to contribute.

I don't agree with the solution in this PR. Maybe the linked issue isn't really a bug at all, it's just a surprising fact.  If there is a rounding error in `capacity()`, we should fix that instead, or we can just let this behavior slide.

**Comment by brson** [comment]
I'm also hesitant about this solution, but also the current state of capacity not being enough for len - that seems like an invariant people would reasonably expect.

**Comment by ranma42** [comment]
The exact computation is:
```Rust
    /// An inverse of `min_capacity`.                                                                                                                                                        
    #[inline]
    fn usable_capacity(&self, cap: usize) -> usize {
        // As the number of entries approaches usable capacity,                                                                                                                                             
        // min_capacity(size) must be smaller than the internal capacity,                                                                                                                                   
        // so that the map is not resized:                                                                                                                                                                  
        // `min_capacity(usable_capacity(x)) <= x`.                                                                                                                                                         
        // The left-hand side can only be smaller due to flooring by integer                                                                                                                                
        // division.                                                                                                                                                                                        
        //                                                                                                                                                                                                  
        // This doesn't have to be checked for overflow since allocation size                                                                                                                               
        // in bytes will overflow earlier than multiplication by 10.                                                                                                                                        
        (cap * 10 + 10 - 1) / 11
    }
```

In general, if `min_capacity` is `usable_size * num / den`, `usable_capacity` should be `(cap * den + den - 1) / num`.

**Comment by brson** [comment]
On second thought this doesn't look so bad to me. It is kinda a hack to paper over a deficiency, but it makes the capacity behave how you expect.

cc @rust-lang/libs

**Comment by alexcrichton** [comment]
I would (perhaps naively) expect `capacity() >= len()` to always hold as well, so I'd be fine with fixing this problem somehow, although I admit to not understanding the underlying intricacies. Maybe @Gankro knows more?

**Comment by Gankro** [comment]
TL;DR it does this right now because of integer division. In particular the actual check we do internally is something like `allocated_space < new_size * x / y` ("is the allocated space 110% of the requested space"), but this function yields `allocated_space * y / x` ("91% of the allocated space"). In my head, these two can't be synched up because of overflow or perf, but I might be mistaken? Relevant code is `HashMap::reserve`. @pczarn might recall the details better.

This was deemed "fine" because the capacity is "I definitely *won't* allocate if you don't exceed this", and we don't actually make any guarantees that we need to reallocate if it is exceeded. It just so happens that HashMap is the only collection where this comes up. Vec is precise, and VecDeque's compensation is `cap - 1`. Everything else is node-based.

I'm fine with the current situation, and I'd be fine with this fix.

**Comment by ranma42** [comment]
@Gankro I believe that the two have not been synched up to keep the code simpler, but in my previous comment I showed how `usable_capacity` should be changed to have the exact rounding.

If the implementation is updated with that code, HashMap behaves just like the other data structures, i.e. its capacity becomes exact: HashMap will not allocate if you insert at most `len() - capacity()` items, it will allocate if you insert `len() - capacity() + 1`.

**Comment by pczarn** [comment]
@ranma42 is right. Adding 1 to `min_capacity` works as well.

**Comment by rthomas** [comment]
Sorry for the lag. I've made the changes that @ranma42 suggested and improved the test for it.

File: src/libstd/collections/hash/map.rs [hunk 1]
```diff
@@ -2418,4 +2421,29 @@ mod test_map {
...
     }
+
+    #[test]
+    fn test_capacity_not_less_than_len() {
+        let mut a = HashMap::new();
+        let mut item = 0;
+
+        for _ in 0..116 {
+            a.insert(item, 0);
+            item = item + 1;
```
Comment (Gankro) [inline src/libstd/collections/hash/map.rs]: `item += 1`

Comment (rthomas) [inline src/libstd/collections/hash/map.rs]: Fixed.

**Comment by Gankro** [comment]
Alright, sorry for the delay. r=me with commits squashed.

(I do wonder if this test is perhaps over-specifying... but if we opt to break it we can of course easily change the test)

**Comment by Gankro** [comment]
@bors r+

Thanks a bunch! :smile_cat:
